crates/sol-macro/src/expand/udt.rs

```diff
@@ -1,32 +1,120 @@
+        );
+
+        #[automatically_derived]
+        impl ::alloy_sol_types::private::SolTypeValue<#name> for <#underlying as ::alloy_sol_types::SolType>::RustType {
+            #[inline]
+            fn stv_to_tokens(&self) -> <#underlying as ::alloy_sol_types::SolType>::Token<'_> {
+                ::alloy_sol_types::private::SolTypeValue::<#underlying>::stv_to_tokens(self)
+            }
+
+            #[inline]
+            fn stv_eip712_data_word(&self) -> ::alloy_sol_types::Word {
+                <#underlying as ::alloy_sol_types::SolType>::tokenize(self).0
+            }
+
+            #[inline]
+            fn stv_abi_encode_packed_to(&self, out: &mut ::alloy_sol_types::private::Vec<u8>) {
+                <#underlying as ::alloy_sol_types::SolType>::abi_encode_packed_to(self, out)
+            }
+        }
+
+        #[automatically_derived]
+        impl #name {
+            /// The Solidity type name.
+            pub const NAME: &'static str = stringify!(@name);
+
+            /// Convert from the underlying value type.
+            #[inline]
+            pub const fn from(value: <#underlying as ::alloy_sol_types::SolType>::RustType) -> Self {
+                Self(value)
+            }
+
+            /// Return the underlying value.
+            #[inline]
+            pub const fn into(self) -> <#underlying as ::alloy_sol_types::SolType>::RustType {
+                self.0
+            }
+
+            /// Return the single encoding of this value, delegating to the
+            /// underlying type.
+            #[inline]
+            pub fn abi_encode(&self) -> ::alloy_sol_types::private::Vec<u8> {
+                <Self as ::alloy_sol_types::SolType>::abi_encode(&self.0)
+            }
+
+            /// Return the packed encoding of this value, delegating to the
+            /// underlying type.
+            #[inline]
+            pub fn abi_encode_packed(&self) -> ::alloy_sol_types::private::Vec<u8> {
+                <Self as ::alloy_sol_types::SolType>::abi_encode_packed(&self.0)
+            }
+        }
+
+        #[automatically_derived]
+        impl ::alloy_sol_types::SolType for #name {
+            type RustType = <#underlying as ::alloy_sol_types::SolType>::RustType;
+            type Token<'a> = <#underlying as ::alloy_sol_types::SolType>::Token<'a>;
+
+            const ENCODED_SIZE: Option<usize> = <#underlying as ::alloy_sol_types::SolType>::ENCODED_SIZE;
+
+            #[inline]
+            fn sol_type_name() -> ::alloy_sol_types::private::Cow<'static, str> {
+                Self::NAME.into()
+            }
+
+            #[inline]
+            fn valid_token(token: &Self::Token<'_>) -> bool {
+                Self::type_check(token).is_ok()
+            }
+
+            #[inline]
+            fn type_check(token: &Self::Token<'_>) -> ::alloy_sol_types::Result<()> {
+                <#underlying as ::alloy_sol_types::SolType>::type_check(token)?;
+                ::alloy_sol_types::private::just_ok(token)
```

Comment:
If you don't mind, I think we can keep the user input with a `#[sol(type_check = "function_name")]` attribute. Otherwise I'll open an issue after.

Reply:
ok, should be done here https://github.com/alloy-rs/core/pull/471/commits/e7e7b8f106e22d68d18ba56f9cd433404c6b410e PTAL